### PR TITLE
[JENKINS-42141] Fix performance issue in code merging Tool installer list

### PR DESCRIPTION
**Pull request by ydubreuil**
[JENKINS-42141](https://issues.jenkins-ci.org/browse/JENKINS-42141)

This patch fixes the performances issue by using a `HashSet` to collect already merged tool installers.

Amends #1972.

@reviewbybees

### CODE REVIEW

**Comment by aheritier** [comment]
cc @varmenise

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java [hunk 1]
```diff
@@ -169,42 +170,23 @@ private boolean isDefaultSchema(List<JSONObject> jsonList) {
             return false;
         }
 
+        /**
+         * Merge a list of ToolInstallerList and removes duplicate tool installers (ie having the same id)
+         * @param jsonList
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Please add the param description to make Javadoc happy

**Review by oleg-nenashev — APPROVED** [review]
🐝 Looks good to me. I had the comment in #1972 regarding it (https://github.com/jenkinsci/jenkins/pull/1972#r50386479 ), but after discussion I agreed with the change since it didn't seem to be a serious issue for small lists.

**Comment by oleg-nenashev** [comment]
@daniel-beck Follow-up to the IRC conversation. From what I see it is covered by tests contributed in https://github.com/jenkinsci/jenkins/pull/1972

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java (hunk 1 above)
Comment (ydubreuil) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Oops, I missed that. Do you mind if I push force the one liner change?

Comment (oleg-nenashev) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Fine for me

**Comment by ydubreuil** [comment]
Like mentioned by @oleg-nenashev , tests are already existing and were added as part of #1972. I didn't bother to write a micro benchmark to prove the efficiency though.

I'm also a bit concerned by the way [JDK are handled](https://github.com/ydubreuil/jenkins/blob/c54bf4eb94ae3806af18167789d7226e5abdc10b/core/src/main/java/hudson/tools/JDKInstaller.java#L818), as the algorithm is pretty similar. I didn't see a performance issue on DEV@cloud but it might have been hidden by the problem we're fixing here.

**Review by stephenc — APPROVED** [review]

**Pull request merged**